cirq/circuits/frozen_circuit.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An immutable version of the Circuit data structure with unassigned qubits.
+
+FrozenCircuits contain a tuple of pseudo-Moments, each of which contains a set
+of Gate + qubit pairs. When the FrozenCircuit is applied to a set of qubits,
+its Gates will be assigned to the qubits as defined by the user.
+"""
+
+from typing import (TYPE_CHECKING, Any, Callable, Iterable, Sequence, Tuple,
+                    overload)
+
+from cirq import devices, ops
+from cirq.circuits import AbstractCircuit, Circuit
+from cirq.circuits.insert_strategy import InsertStrategy
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class FrozenCircuit(AbstractCircuit):
+
+    def __init__(self,
+                 *contents: 'cirq.OP_TREE',
+                 strategy: 'cirq.InsertStrategy' = InsertStrategy.EARLIEST,
+                 device: 'cirq.Device' = devices.UNCONSTRAINED_DEVICE) -> None:
+        base = Circuit(contents, strategy=strategy, device=device)
+        self._moments = tuple(base.moments)
+        self._device = base.device
+
+        # These variables are memoized when first requested.
+        self._num_qubits = None
+        self._qid_shape = None
+        self._all_qubits = None
+        self._all_operations = None
+
+    @property
+    def moments(self) -> Sequence['cirq.Moment']:
+        return self._moments
+
+    @property
+    def device(self) -> devices.Device:
+        return self._device
+
+    @staticmethod
+    def freeze(circuit: Circuit) -> 'FrozenCircuit':
```

Comment:
Instead of a static method I'd suggest having a `freeze` method on `AbtractCircuit` that returns `FrozenCircuit`. This would create a new frozen circuit if called on a mutable be circuit, and return `self` if called on a frozen circuit.